File: frontend/src/lib/chat/permission-selection.ts
import { PERMISSION_METADATA, type PermissionMetadata } from '$lib/generated/permissions';

export function permissionMask(value: string): bigint {
  const normalized = String(value).trim();
  if (!/^\d+$/.test(normalized)) {
    throw new Error('Permissions must be a non-negative whole number.');
  }
  return BigInt(normalized);
}

export function permissionSelected(value: string, bit: bigint): boolean {
  return Boolean(permissionMask(value) & bit);
}

export function setPermissionSelected(value: string, bit: bigint, selected: boolean): string {
  const current = permissionMask(value);
  return (selected ? current | bit : current & ~bit).toString();
}

export function selectedPermissionMetadata(value: string): PermissionMetadata[] {
  const current = permissionMask(value);
  return PERMISSION_METADATA.filter((item) => Boolean(current & item.bit));
}

File: mobile/lib/src/domain/permission_selection.dart
import 'package:kaede_mobile/src/protocol/generated.dart';

BigInt applicationPermissionMask(String value) {
  final normalized = value.trim();
  if (!RegExp(r'^\d+$').hasMatch(normalized)) {
    throw const FormatException(
      'Permissions must be a non-negative whole number.',
    );
  }
  return BigInt.parse(normalized);
}

bool applicationPermissionSelected(BigInt value, PermissionMetadata item) =>
    value & BigInt.from(item.bit) != BigInt.zero;

BigInt setApplicationPermission(
  BigInt value,
  PermissionMetadata item,
  bool selected,
) {
  final bit = BigInt.from(item.bit);
  return selected ? value | bit : value & ~bit;
}

List<PermissionMetadata> get applicationInstallPermissions => permissionMetadata
    .where((item) =>
        item.resourceScopes.contains('guild') ||
        item.resourceScopes.contains('channel'))
    .toList(growable: false);

List<PermissionMetadata> selectedApplicationPermissions(String value) {
  final mask = applicationPermissionMask(value);
  return applicationInstallPermissions
      .where((item) => applicationPermissionSelected(mask, item))
      .toList(growable: false);
}
